airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -170,14 +171,12 @@ def _get_concurrent_state(self, state: MutableMapping[str, Any]) -> Tuple[Cursor
         return self._connector_state_converter.convert_from_sequential_state(self._cursor_field, state, self._start)
 
     def observe(self, record: Record) -> None:
-        if self._slice_boundary_fields:
-            # Given that slicing is done using the cursor field, we don't need to observe the record as we assume slices will describe what
-            # has been emitted. Assuming there is a chance that records might not be yet populated for the most recent slice, use a lookback
-            # window
-            return
+        most_recent_record = self._most_recent_record_per_partition.get(record.partition.stream_name())
```

Comment:
The name of the variable says `_per_partition` but the key with which we access the most recent records is `partition.stream_name`. If it is indeed the stream name we are using, it feels like we don't need a dict as we instantiate one cursor per stream (at least, it might be more when we will support per partition)

Reply:
It should be pre-partition, due to the asynchronous nature of partition processing. We use `most_recent_cursor_value` only for the last slice to limit it from the right side. However, this slice can only be definitively identified as the last slice after all slices have been processed. During processing, an error might occur, preventing us from closing our last slice. Therefore, we need to ensure that at every moment during processing, we have our `most_recent_cursor_value` available for each partition.